storage/sealer/sched_resources.go

```diff
@@ -29,43 +31,52 @@ type taskCounter struct {
 
 func newTaskCounter() *taskCounter {
 	return &taskCounter{
-		taskCounters: map[sealtasks.SealTaskType]int{},
+		taskCounters: make(map[sealtasks.SealTaskType]map[uuid.UUID]bool),
 	}
 }
 
-func (tc *taskCounter) Add(tt sealtasks.SealTaskType) {
+func (tc *taskCounter) Add(tt sealtasks.SealTaskType, schedID uuid.UUID) {
 	tc.lk.Lock()
 	defer tc.lk.Unlock()
-	tc.taskCounters[tt]++
+	tc.getUnlocked(tt)[schedID] = true
 }
 
-func (tc *taskCounter) Free(tt sealtasks.SealTaskType) {
+func (tc *taskCounter) Free(tt sealtasks.SealTaskType, schedID uuid.UUID) {
 	tc.lk.Lock()
 	defer tc.lk.Unlock()
-	tc.taskCounters[tt]--
+	delete(tc.getUnlocked(tt), schedID)
```

Comment:
So I believe this has a bug which makes it not work as we'd expect:
* First we call .Add for preparing
* Then we call .Add on active in .withResources
* In .withResources we call .Free on preparing
* Now the task runs - and we don't count the resources
* Then we call .Free on preparing

This can be fixed by swapping `map[sealtasks.SealTaskType]map[uuid.UUID]bool` for `map[sealtasks.SealTaskType]map[uuid.UUID]int` and count how many times Add/Free was called.

Reply:
Yeah its allowing more than it should as-is.
Trying that and counting +1/-1 for each Add/Free invocation leads to wrong stuck behavior again:
```
ID        Sector  Worker    Hostname    Task  State        Time
00000000  23      4154622e  x  GET   assigned(1)  21.2s
00000000  24      4154622e  x  GET   assigned(2)  21.2s
```
Thats interesting, since only reasonable idea I have is that there's an invocation missing somewhere.

Reply:
@magik6k 
Apparently the issue is the last Free for a SchedId happens after the last `WINDOW assignPreparingWork` 
(which in turn says ` not scheduling on worker  for startPreparing`).
A quick test of returning `update=true` in `schedWorker::waitForUpdates` to force a regular invocation of that leads to the "GET backlog" getting fully resolved over time as well.

Reply:
@magik6k Any update on this? _This is now broken in 1.21, 1.22 and 1.23_.

Reply:
> only reasonable idea I have is that there's an invocation missing somewhere.

How did you implement `Free` with the counter? Did you `delete` from the map when the counter reached zero? If not the `tasks := a.taskCounters.Get(tt) [...] if len(tasks) >= needRes.MaxConcurrent` below would not work correctly as it still counted the zero enties.

Reply:
@magik6k Yeah I've accounted for that:
https://github.com/steffengy/lotus/commit/a3b7ec2447af0849b10a3e65fe67120a3f8781ff
(I think else it also wouldnt work with making "scheduler entry" more often by returning `update=true` in `schedWorker::waitForUpdates` as that just evaluates the same at a later time; so its a timing/ordering issue last Free happening after the next scheduler entry for startPreparing and stuck there until next scheduling happening)

Reply:
I've spent some time playing with it (in https://github.com/filecoin-project/lotus/pull/10850), and I think I found the issue with integer counters - https://github.com/filecoin-project/lotus/pull/10850/commits/240fc6d3b905f69ec61bebcfbe18c3f1cd7686a6

Would be great if you could see if it works for you

Reply:
Great! Yeah in a quick test that looks working and fix-wise it seems plausible that that was the issue.
Closing this PR in favor of yours.